### PR TITLE
backup,restore: support prefix-free backups

### PR DESCRIPTION
**Pull request by dt**
This extends 24.1 backups to be produced with the tenant ID and table ID prefix on every key stored in the backup elided from the stored keys. The elided prefix can be derived from the prefix of the span being restored so it can be elided from the individual keys with no loss of information.

### CODE REVIEW

**Comment by dt** [comment]
I think we'll need https://github.com/cockroachdb/cockroach/pull/116902 first to make the online restore tests pass

**Review by stevendanna — COMMENTED** [review]
:lgtm: Overall.

Pretty impressive that all the tests (other than the ones we absolutely knew wouldn't work) "just pass."

As discussed in chat, one test that would be good to do is to make sure the logic around _not_ eliding prefixes works by making sure we run the mixed-version roachtest on this.

File: pkg/ccl/backupccl/file_sst_sink.go
Comment (msbutler) [inline pkg/ccl/backupccl/file_sst_sink.go:225]: could you add a test case to file_sink_test.go that validates that two's sst's with different prefixes get flushed to different ssts?

Comment (msbutler) [reply pkg/ccl/backupccl/file_sst_sink.go:225]: it may also be worth adding an outer loop to that that so that all test cases deal with/without stripping prefixes.

File: pkg/ccl/backupccl/backup_job.go
Comment (dt) [reply pkg/ccl/backupccl/backup_job.go]: Done.

File: pkg/ccl/backupccl/restore_online.go
Comment (stevendanna) [inline pkg/ccl/backupccl/restore_online.go:131]: [nit] extra space. But leave it if you like it :D.

File: pkg/ccl/backupccl/restore_online.go
Comment (stevendanna) [inline pkg/ccl/backupccl/restore_online.go:128]: There is a tenant rekey bit below, so mind putting a few more words here about what still needs to be handled?

File: pkg/ccl/backupccl/restore_planning.go
Comment (stevendanna) [inline pkg/ccl/backupccl/restore_planning.go:2086]: 🎉 I think we can probably remove the `checkRewritesAreNoops` function for now as well?

**Review by stevendanna — COMMENTED** [review]
Thanks for adding that option.

I left a couple of nits but then I stopped since most were just making the code different and not better. Feel free to ignore.

In the old restore path, having to check our un-rewritten key against the endKey is a little bit of a bummer just because it _seems_ like a good amount of work could be avoided if we didn't have to do that.  But, onward to online restore.

I'll let @msbutler approve one the additional test question is resolved.

**Comment by dt** [comment]
@stevendanna I didn't mean to push the version of this branch that I rebased on #116902. I'm gonna undo that rebase so hold off here for now.

**Comment by dt** [comment]
Okay, un-rebased to remove accidental inclusion of https://github.com/cockroachdb/cockroach/pull/116902 and reworked file_sst_sinks_test to include table IDs and support eliding. RFAL now.

File: pkg/ccl/backupccl/file_sst_sink.go
Comment (dt) [reply pkg/ccl/backupccl/file_sst_sink.go:225]: Done

**Review by msbutler — APPROVED** [review]

**Pull request merged**
